### ISSUE TITLE
Token Auth / Stateless / DB Lookup

### ISSUE THREAD

**Issue by pfazzino** [tymondesigns/jwt-auth#183]
Just some general comments regarding this implementation, helpful and excellent as it is. 

Although the JWT spec makes no claims in terms of statelessness or what you actually do with the token on receipt in any real detail, one of the claimed advantages of a token based approach compared with session based is the fact that the token is stateless, able to be easily verified on any endpoint without the baggage and requirements associated with sessions (session storage, load balancing challenges, etc...).

One concern I have with the jwt-auth lib is that part of it's authentication process is to trigger an identity lookup via the Laravel authentication process, this will usually involve a db query to hydrate the model regardless of whether it's actually required or not. If you're going to be triggering a db lookup to authenticate a user associated with a jwt token you are losing one of the benefits imho. You could just as easily move session storage to a central location and gain/lose the same set of benefits/challenges.

For our own implementation which leverages jwt-auth, we have middleware which simply verifies the token from a payload/signature perspective but doesn't trigger the authentication flow. This is defered to when we actually need the User object, e.g. Auth::user() via implementation of custom UserProvider/Guard implementation and overriding of the user()/check()/id() functions, or when we need to do a token refresh which does a full auth flow, checks account is valid, not disabled, etc...

The advantage of this approach (and one of the selling points of tokens/jwt imho) is that authentication is a very lightweight process which simply validates the incoming token. The disadvantage is that any change in the user model (e.g. account disabled) is not picked up until a refresh/re-auth cycle. This becomes a balancing act of token ttl vs processing overhead.

In summary, imho it feels like jwt-auth should not be triggering the User model hydration as part of it's standard middleware token validation and should provide explicit separation of these functions. Sure you can write custom middleware as we have but I'm guessing a lot of users will use the out of the box offerings and lose some of the benefits of jwt/tokens which would beg the question why not stick with sessions in the first place.

Interested to hear other's thoughts on this topic...

**Comment by dani3l** [comment]
Actually I think that the out-of-the-box middleware is pefrect for the vast majority of developers
If someone need to just validate the token, he can create a custom middleware like you did.

I, for example, created a custom middleware to check if user changed login credentials (user_model.credentials_updated_at >= payload.iat)

**Comment by tymondesigns** [comment]
Sorry, forgot to reply to this.

I understand where your coming from here. However I do feel that it's probably beyond the scope of the package to offer such, arguably, bespoke functionality.

I am re-thinking a few things at the moment, one being out of the box middlewares. So may well figure something out in regards to this (no promises :smile:)

I have also added some little helper methods that should aid in implementing the functionality you mentioned, such as [these](https://github.com/tymondesigns/jwt-auth/blob/develop/src/JWTAuth.php#L128-L156)

As always, PR's are welcomed if you feel something is missing or would improve how things work

**Comment by biblicalph** [comment]
@pfazzino I am not sure how you are verifying the payload but I believe yes verifying the signature adheres to the JWT standard. In my own case however, I find that half the time, requests involve looking up some records from the DB which means authentication flow is required nonetheless. 
I will be however grateful if you can share some code from your implementation - will be useful in strictly API terms where a request might be to simply fetch data in which case jwt-auth authentication of tokens is not entirely valid. 
Regards

**Comment by pfazzino** [comment]
I'm away at present and won't have access to the code for a couple of weeks
but the general approach we're using is to use jwt-auth in custom
middleware to just parse/verify the jwt payload, not trigger the whole auth
via the user lookup by id which normally occurs. This means no store lookup
is required to authenticate requests.

We have a custom auth provider which also overrides check/user functions
which now use the stored user id initially parsed from the token to then
make the dB lookup on request. Where the token sub/id is stored is
completely up to you, it just needs to be somewhere the auth provider can
find later on if needed.

The above means user model hydration is only done on demand and is
completely separate from auth.

If you're still after code I'm happy to share on my return, just let me
know and I'll post something up.

Regards

Paul

**Comment by biblicalph** [comment]
@pfazzino Thanks for the prompt response. I am quite new to JWT and get the part about authenticating the signature but not so much the part about authenticating the payload. 
So yes, I'll look forward to some code on your return. Have fun. 

Regards, 
KB
